dkg/actions_active.go

```diff
@@ -0,0 +1,443 @@
+		// we send the proposal to joiners and remainers and error if they don't respond
+		err = d.network.Send(
+			me,
+			util.Concat(nextState.Joining, nextState.Remaining),
+			func(client net.DKGClient, peer net.Peer) (*drand.EmptyResponse, error) {
+				return client.Propose(ctx, peer, &terms)
+			},
+		)
+		if err != nil {
+			return err
+		}
+
+		// we make a best effort attempt to send the proposal to the leaver, but if their node is e.g. turned off then
+		// we ignore the error
+		if len(nextState.Leaving) > 0 {
+			err = d.network.Send(me, nextState.Leaving, func(client net.DKGClient, peer net.Peer) (*drand.EmptyResponse, error) {
+				return client.Propose(ctx, peer, &terms)
+			})
+
+			if err != nil {
+				d.log.Warnw("could not send proposal to a leaving participant", "err", err)
+			}
+		}
+
+		if err := d.store.SaveCurrent(beaconID, nextState); err != nil {
+			return err
+		}
+
+		d.log.Infow("Finished proposing a new DKG", "beaconID", beaconID)
+		return nil
+	}
+
+	// if there's an error sending to a party or saving the state, attempt a rollback by issuing an abort
+	rollback := func(err error) {
+		allParticipants := util.Concat(nextState.Joining, nextState.Remaining, nextState.Leaving)
+		d.log.Errorw("There was an error proposing a DKG", "err", err, "beaconID", beaconID)
+		_ = d.attemptAbort(ctx, me, allParticipants, beaconID)
+	}
+
+	return responseOrError(rollbackOnError(sendProposalToAllAndStoreState, rollback))
+}
+
+func (d *DKGProcess) StartAbort(ctx context.Context, options *drand.AbortOptions) (*drand.EmptyResponse, error) {
+	beaconID := options.BeaconID
+	d.log.Infow("Aborting DKG", "beaconID", beaconID)
+
+	me, err := d.identityForBeacon(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	current, err := d.store.GetCurrent(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	if !util.EqualParticipant(current.Leader, me) {
+		return nil, errors.New("cannot abort the DKG if you aren't the leader")
+	}
+
+	nextState, err := current.Aborted()
+	if err != nil {
+		return nil, err
+	}
+
+	allParticipants := util.Concat(nextState.Joining, nextState.Remaining, nextState.Leaving)
+	if err := d.attemptAbort(ctx, me, allParticipants, beaconID); err != nil {
+		return nil, err
+	}
+
+	if err := d.store.SaveCurrent(beaconID, nextState); err != nil {
+		d.log.Errorw("error aborting the DKG", "error", err, "beaconID", beaconID)
+	} else {
+		d.log.Infow("DKG aborted successfully", "beaconID", beaconID)
+	}
+	return responseOrError(err)
+}
+
+func (d *DKGProcess) StartExecute(ctx context.Context, options *drand.ExecutionOptions) (*drand.EmptyResponse, error) {
+	beaconID := options.BeaconID
+
+	stateTransition := func(me *drand.Participant, current *DBState) (*DBState, error) {
+		if !util.EqualParticipant(current.Leader, me) {
+			return nil, errors.New("cannot start execution if you aren't the leader")
+		}
+		return current.Executing(me)
+	}
+
+	callback := func(me *drand.Participant, nextState *DBState) error {
```

Comment:
I admit that I'm lost - what is this callback doing really. Just calling "execute" each time there's ... a new "state" ? Isn't then the state transition function ? Why should be this be a callback that we can change at ease and not a well defined function  that we can lookup somewhere and say "this is how we transition from each state"